Simplify observers declaration with functools.partial

dzhulgakov opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#26492 Simplify observers declaration with functools.partial**

Previous definition of observers was quite clumsy - with things like `default_observer()()`. This PR strips a way a lot of craft and allows to pass just class names directly. In order to override default arguments either `functools.partial` can be used or convenient wrapper `MyObserver.with_args(x=1)` is provided.

Also rename `QConfig_dynamic` to `QConfigDynamic` because it violates the naming convention.

Differential Revision: [D17521265](https://our.internmc.facebook.com/intern/diff/D17521265)

Commit:
Simplify observers declaration with functools.partial

[ghstack-poisoned]

Commit:
Update on "Simplify observers declaration with functools.partial"




[ghstack-poisoned]

Commit:
Update on "Simplify observers declaration with functools.partial"




[ghstack-poisoned]

Commit:
Update on "Simplify observers declaration with functools.partial"


Previous definition of observers was quite clumsy - with things like `default_observer()()`. This PR strips a way a lot of craft and allows to pass just class names directly. In order to override default arguments either `functools.partial` can be used or convenient wrapper `MyObserver.with_args(x=1)` is provided.

Also rename `QConfig_dynamic` to `QConfigDynamic` because it violates the naming convention.

[ghstack-poisoned]

jerryzh168 commented on torch/quantization/QConfig.py:
```diff
@@ -2,19 +2,60 @@
 
-default_qconfig = QConfig(activation=default_observer(),
-                          weight=default_weight_observer())
+    Note that QConfig needs to contain observer **classes** (like MinMaxObserver), not the concrete instances.
```
I think it's either **classes** or **functools.partial object**?

jerryzh168 replied:
or we can say, it's something that will return an instance when called. sounds a bit confusing though.

jerryzh168 commented on test/common_quantization.py:
```diff
@@ -279,8 +280,8 @@ def __init__(self):
         }
-        custom_qconfig = QConfig(activation=default_observer(**custom_options),
-                                 weight=default_weight_observer())
+        custom_qconfig = QConfig(activation=partial(default_observer, **custom_options),
```
I think we can do `with_args` here?

dzhulgakov replied:
oh, yeah, missed this one, thanks :)

jerryzh168 reviewed (approved):
Looks good Thanks

Commit:
Update on "Simplify observers declaration with functools.partial"


Previous definition of observers was quite clumsy - with things like `default_observer()()`. This PR strips a way a lot of craft and allows to pass just class names directly. In order to override default arguments either `functools.partial` can be used or convenient wrapper `MyObserver.with_args(x=1)` is provided.

Also rename `QConfig_dynamic` to `QConfigDynamic` because it violates the naming convention.

[ghstack-poisoned]

Commit:
Update on "Simplify observers declaration with functools.partial"


Previous definition of observers was quite clumsy - with things like `default_observer()()`. This PR strips a way a lot of craft and allows to pass just class names directly. In order to override default arguments either `functools.partial` can be used or convenient wrapper `MyObserver.with_args(x=1)` is provided.

Also rename `QConfig_dynamic` to `QConfigDynamic` because it violates the naming convention.

[ghstack-poisoned]

Commit:
Update on "Simplify observers declaration with functools.partial"


Previous definition of observers was quite clumsy - with things like `default_observer()()`. This PR strips a way a lot of craft and allows to pass just class names directly. In order to override default arguments either `functools.partial` can be used or convenient wrapper `MyObserver.with_args(x=1)` is provided.

Also rename `QConfig_dynamic` to `QConfigDynamic` because it violates the naming convention.

Differential Revision: [D17521265](https://our.internmc.facebook.com/intern/diff/D17521265)

[ghstack-poisoned]
